File: dashboard/static/charts.js
/**
 * charts.js — Plotly chart generators for the dashboard
 * Each function takes data from the Flask API and renders a chart
 */

const PLOTLY_CFG = { displayModeBar: false, responsive: true };

const PLOTLY_LAYOUT = (overrides = {}) => ({
  margin:         { t: 10, r: 10, b: 40, l: 50 },
  paper_bgcolor:  "rgba(0,0,0,0)",
  plot_bgcolor:   "rgba(0,0,0,0)",
  font: {
    family: "-apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif",
    size: 12, color: "#374151"
  },
  showlegend: false,
  ...overrides,
});

// ─── Format helpers ────────────────────────────────────────────────────
export function fmtTime(sec) {
  if (!sec || sec < 0) return "0m";
  const h = Math.floor(sec / 3600);
  const m = Math.floor((sec % 3600) / 60);
  return h ? `${h}h ${m}m` : `${m}m`;
}

export function fmtNumber(n) {
  if (n >= 1000) return (n / 1000).toFixed(1) + "k";
  return String(n);
}

// ─── Daily activity bar + line chart ──────────────────────────────────
export function renderDailyChart(elementId, rows) {
  if (!rows?.length) return;
  Plotly.newPlot(elementId, [
    {
      type: "bar",
      name: "Visits",
      x: rows.map(r => r.day),
      y: rows.map(r => r.total_visits || 0),
      marker: { color: "#4f46e5", opacity: 0.85 },
    },
    {
      type: "scatter",
      name: "Minutes online",
      x: rows.map(r => r.day),
      y: rows.map(r => Math.round((r.total_time_sec || 0) / 60)),
      mode: "lines+markers",
      yaxis: "y2",
      line: { color: "#10b981", width: 2 },
      marker: { size: 5, color: "#10b981" },
    },
  ],
  PLOTLY_LAYOUT({
    yaxis:  { title: "Visits",   gridcolor: "#f3f4f6", zeroline: false },
    yaxis2: { title: "Minutes",  overlaying: "y", side: "right",
              gridcolor: "#f3f4f6", zeroline: false },
    showlegend: true,
    legend: { x: 0, y: 1.15, orientation: "h" },
    bargap: 0.3,
  }),
  PLOTLY_CFG);
}

// ─── Category donut chart ──────────────────────────────────────────────
export function renderCategoryChart(elementId, rows) {
  if (!rows?.length) return;
  Plotly.newPlot(elementId, [{
    type: "pie",
    labels: rows.map(r => r.name || "Other"),
    values: rows.map(r => r.visits),
    marker: { colors: rows.map(r => r.color_hex || "#888") },
    hole: 0.42,
    textinfo: "label+percent",
    textfont: { size: 11 },
    hovertemplate: "<b>%{label}</b><br>%{value} visits<br>%{percent}<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    showlegend: true,
    legend: { orientation: "v", x: 1.05, y: 0.5 },
    margin: { t: 10, r: 130, b: 10, l: 10 },
  }),
  PLOTLY_CFG);
}

// ─── Top domains horizontal bar ────────────────────────────────────────
export function renderTopSitesChart(elementId, rows) {
  if (!rows?.length) return;
  const top = rows.slice(0, 10);
  Plotly.newPlot(elementId, [{
    type: "bar",
    orientation: "h",
    y: top.map(r => r.domain).reverse(),
    x: top.map(r => r.visits).reverse(),
    marker: { color: "#4f46e5", opacity: 0.8 },
    hovertemplate: "<b>%{y}</b><br>%{x} visits<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    margin: { t: 10, r: 20, b: 30, l: 130 },
    xaxis: { gridcolor: "#f3f4f6", zeroline: false },
  }),
  PLOTLY_CFG);
}

// ─── Search keyword frequency bar ─────────────────────────────────────
export function renderKeywordsChart(elementId, rows) {
  if (!rows?.length) return;
  const top = rows.slice(0, 20);
  Plotly.newPlot(elementId, [{
    type: "bar",
    orientation: "h",
    y: top.map(r => r.keyword).reverse(),
    x: top.map(r => r.frequency).reverse(),
    marker: { color: "#10b981", opacity: 0.85 },
    hovertemplate: "<b>%{y}</b><br>%{x} searches<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    margin: { t: 10, r: 20, b: 30, l: 110 },
    xaxis: { gridcolor: "#f3f4f6", zeroline: false },
  }),
  PLOTLY_CFG);
}

// ─── Search engine pie ─────────────────────────────────────────────────
export function renderEnginesChart(elementId, rows) {
  if (!rows?.length) return;
  Plotly.newPlot(elementId, [{
    type: "pie",
    labels: rows.map(r => r.engine),
    values: rows.map(r => r.total_searches),
    hole: 0.38,
    textinfo: "label+percent",
    marker: {
      colors: ["#4f46e5","#10b981","#f59e0b","#ef4444","#8b5cf6"]
    },
  }],
  PLOTLY_LAYOUT({ margin: { t: 10, r: 10, b: 10, l: 10 } }),
  PLOTLY_CFG);
}

// ─── Social platform time bar ──────────────────────────────────────────
export function renderSocialTimeChart(elementId, rows) {
  if (!rows?.length) return;
  Plotly.newPlot(elementId, [{
    type: "bar",
    x: rows.map(r => r.platform),
    y: rows.map(r => Math.round((r.total_sec || 0) / 60)),
    marker: { color: "#4f46e5", opacity: 0.85 },
    hovertemplate: "<b>%{x}</b><br>%{y} minutes<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    yaxis: { title: "Minutes", gridcolor: "#f3f4f6", zeroline: false },
  }),
  PLOTLY_CFG);
}

// ─── Social actions grouped bar ────────────────────────────────────────
export function renderSocialActionsChart(elementId, rows) {
  if (!rows?.length) return;
  Plotly.newPlot(elementId, [
    {
      type: "bar", name: "Posts",
      x: rows.map(r => r.platform),
      y: rows.map(r => r.posts   || 0),
      marker: { color: "#4f46e5" },
    },
    {
      type: "bar", name: "Likes",
      x: rows.map(r => r.platform),
      y: rows.map(r => r.likes   || 0),
      marker: { color: "#10b981" },
    },
    {
      type: "bar", name: "Scrolls",
      x: rows.map(r => r.platform),
      y: rows.map(r => r.scrolls || 0),
      marker: { color: "#f59e0b" },
    },
    {
      type: "bar", name: "Comments",
      x: rows.map(r => r.platform),
      y: rows.map(r => r.comments || 0),
      marker: { color: "#ef4444" },
    },
  ],
  PLOTLY_LAYOUT({
    barmode: "group", showlegend: true,
    legend: { x: 0, y: 1.15, orientation: "h" },
    yaxis: { gridcolor: "#f3f4f6", zeroline: false },
  }),
  PLOTLY_CFG);
}

// ─── Activity heatmap ──────────────────────────────────────────────────
export function renderHeatmapChart(elementId, rows) {
  if (!rows?.length) return;
  const DAYS  = ["Sun","Mon","Tue","Wed","Thu","Fri","Sat"];
  const HOURS = Array.from({ length: 24 }, (_, i) => `${i}:00`);
  const z     = Array.from({ length: 7 }, () => new Array(24).fill(0));

  rows.forEach(r => {
    const d = parseInt(r.day_of_week), h = parseInt(r.hour_of_day);
    if (d >= 0 && d < 7 && h >= 0 && h < 24) {
      z[d][h] = r.visit_count || 0;
    }
  });

  Plotly.newPlot(elementId, [{
    type: "heatmap",
    z, x: HOURS, y: DAYS,
    colorscale: [[0,"#eef2ff"],[0.5,"#6366f1"],[1,"#312e81"]],
    showscale: true,
    hovertemplate: "<b>%{y} %{x}</b><br>%{z} visits<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    margin: { t: 10, r: 80, b: 60, l: 50 },
    xaxis: { tickangle: -45 },
  }),
  PLOTLY_CFG);
}

// ─── App usage horizontal bar ──────────────────────────────────────────
export function renderAppUsageChart(elementId, rows) {
  if (!rows?.length) return;
  const top = rows.slice(0, 10);
  Plotly.newPlot(elementId, [{
    type: "bar",
    orientation: "h",
    y: top.map(r => r.app_name).reverse(),
    x: top.map(r => Math.round((r.total_sec || 0) / 60)).reverse(),
    marker: { color: "#8b5cf6", opacity: 0.85 },
    hovertemplate: "<b>%{y}</b><br>%{x} minutes<extra></extra>",
  }],
  PLOTLY_LAYOUT({
    margin: { t: 10, r: 20, b: 30, l: 120 },
    xaxis: { title: "Minutes", gridcolor: "#f3f4f6", zeroline: false },
  }),
  PLOTLY_CFG);
}

File: dashboard/charts.py
"""
charts.py — Server-side Plotly chart generators (optional)
Use these if you want to render charts as static images or JSON on the server.
The dashboard.html uses client-side Plotly instead, but these helpers
are useful for exporting reports or generating chart data via the API.
"""

import json


def _base_layout(title="", **kwargs):
    layout = {
        "margin":        {"t": 40, "r": 20, "b": 50, "l": 60},
        "paper_bgcolor": "white",
        "plot_bgcolor":  "#fafafa",
        "font":          {"family": "Arial, sans-serif", "size": 12},
        "showlegend":    False,
    }
    if title:
        layout["title"] = {"text": title, "font": {"size": 16}}
    layout.update(kwargs)
    return layout


def daily_activity_chart(rows):
    """Bar + line chart showing daily visits and time online."""
    days         = [str(r["day"]) for r in rows]
    visits       = [r.get("total_visits", 0) for r in rows]
    minutes      = [round((r.get("total_time_sec") or 0) / 60) for r in rows]

    return {
        "data": [
            {
                "type": "bar", "name": "Visits",
                "x": days, "y": visits,
                "marker": {"color": "#4f46e5", "opacity": 0.85},
            },
            {
                "type": "scatter", "name": "Minutes online",
                "x": days, "y": minutes,
                "mode": "lines+markers", "yaxis": "y2",
                "line": {"color": "#10b981", "width": 2},
                "marker": {"size": 5, "color": "#10b981"},
            },
        ],
        "layout": _base_layout(
            title="Daily Activity",
            yaxis={"title": "Visits", "gridcolor": "#e5e7eb"},
            yaxis2={"title": "Minutes", "overlaying": "y",
                    "side": "right", "gridcolor": "#e5e7eb"},
            showlegend=True,
            bargap=0.3,
        ),
    }


def category_pie_chart(rows):
    """Donut chart showing visit distribution by category."""
    return {
        "data": [{
            "type": "pie",
            "labels": [r.get("name", "Other") for r in rows],
            "values": [r.get("visits", 0) for r in rows],
            "marker": {"colors": [r.get("color_hex", "#888") for r in rows]},
            "hole": 0.42,
            "textinfo": "label+percent",
        }],
        "layout": _base_layout(title="Category Breakdown"),
    }


def top_domains_chart(rows, limit=10):
    """Horizontal bar chart of top visited domains."""
    top     = rows[:limit]
    domains = [r.get("domain", "") for r in reversed(top)]
    visits  = [r.get("visits", 0) for r in reversed(top)]

    return {
        "data": [{
            "type": "bar", "orientation": "h",
            "y": domains, "x": visits,
            "marker": {"color": "#4f46e5", "opacity": 0.8},
        }],
        "layout": _base_layout(
            title="Top Domains",
            margin={"t": 40, "r": 20, "b": 40, "l": 140},
            xaxis={"gridcolor": "#e5e7eb"},
        ),
    }


def heatmap_chart(rows):
    """Activity heatmap: day of week × hour of day."""
    DAYS  = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
    HOURS = [f"{h}:00" for h in range(24)]
    z     = [[0] * 24 for _ in range(7)]

    for r in rows:
        d = int(r.get("day_of_week", 0))
        h = int(r.get("hour_of_day", 0))
        if 0 <= d < 7 and 0 <= h < 24:
            z[d][h] = r.get("visit_count", 0)

    return {
        "data": [{
            "type": "heatmap",
            "z": z, "x": HOURS, "y": DAYS,
            "colorscale": [[0, "#eef2ff"], [0.5, "#6366f1"], [1, "#312e81"]],
            "showscale": True,
        }],
        "layout": _base_layout(
            title="Activity Heatmap",
            margin={"t": 40, "r": 80, "b": 60, "l": 50},
            xaxis={"tickangle": -45},
        ),
    }


def keyword_frequency_chart(rows, limit=20):
    """Horizontal bar showing most searched keywords."""
    top      = rows[:limit]
    keywords = [r.get("keyword", "") for r in reversed(top)]
    freqs    = [r.get("frequency", 0) for r in reversed(top)]

    return {
        "data": [{
            "type": "bar", "orientation": "h",
            "y": keywords, "x": freqs,
            "marker": {"color": "#10b981", "opacity": 0.85},
        }],
        "layout": _base_layout(
            title="Top Search Keywords",
            margin={"t": 40, "r": 20, "b": 40, "l": 110},
            xaxis={"gridcolor": "#e5e7eb"},
        ),
    }


def social_platform_chart(rows):
    """Grouped bar chart for social platform actions."""
    platforms = [r.get("platform", "") for r in rows]

    return {
        "data": [
            {
                "type": "bar", "name": "Posts",
                "x": platforms, "y": [r.get("posts", 0) for r in rows],
                "marker": {"color": "#4f46e5"},
            },
            {
                "type": "bar", "name": "Likes",
                "x": platforms, "y": [r.get("likes", 0) for r in rows],
                "marker": {"color": "#10b981"},
            },
            {
                "type": "bar", "name": "Scrolls",
                "x": platforms, "y": [r.get("scrolls", 0) for r in rows],
                "marker": {"color": "#f59e0b"},
            },
        ],
        "layout": _base_layout(
            title="Social Media Actions",
            barmode="group",
            showlegend=True,
            yaxis={"gridcolor": "#e5e7eb"},
        ),
    }


def app_usage_chart(rows, limit=10):
    """Horizontal bar of top app usage by minutes."""
    top      = rows[:limit]
    apps     = [r.get("app_name", "") for r in reversed(top)]
    minutes  = [round((r.get("total_sec") or 0) / 60) for r in reversed(top)]

    return {
        "data": [{
            "type": "bar", "orientation": "h",
            "y": apps, "x": minutes,
            "marker": {"color": "#8b5cf6", "opacity": 0.85},
        }],
        "layout": _base_layout(
            title="App Usage (minutes)",
            margin={"t": 40, "r": 20, "b": 40, "l": 130},
            xaxis={"title": "Minutes", "gridcolor": "#e5e7eb"},
        ),
    }
